src/psyclone/tests/dependency_test.py

```diff
@@ -0,0 +1,358 @@
+    ''' A variable on the LHS of an assignment must only occur once,
+    which is a restriction of PSyclone.
+
+    '''
+    reader = FortranStringReader('''program test_prog
+                                 g(g(1)) = 1
+                                 end program test_prog''')
+    ast = parser(reader)
+    psy = PSyFactory(API).create(ast)
+    schedule = psy.invokes.get("test_prog").schedule
+
+    indirect_addressing = schedule[0]
+    assert isinstance(indirect_addressing, Assignment)
+    var_accesses = VariablesAccessInfo()
+    from psyclone.parse.utils import ParseError
+    with pytest.raises(ParseError) as err:
+        indirect_addressing.reference_accesses(var_accesses)
+    assert "The variable 'g' appears more than once on the left-hand side "\
+           "of an assignment." in str(err)
+
+
+def test_if_statement(parser):
+    ''' Tests handling an if statement
+    '''
+    reader = FortranStringReader('''program test_prog
+                                 if (a .eq. b) then
+                                    p(i) = q(i)
+                                 else
+                                   q(i) = r(i)
+                                 endif
+                                 end program test_prog''')
+    ast = parser(reader)
+    psy = PSyFactory(API).create(ast)
+    schedule = psy.invokes.get("test_prog").schedule
+
+    if_stmt = schedule.children[0]
+    assert isinstance(if_stmt, IfBlock)
+    var_accesses = VariablesAccessInfo()
+    if_stmt.reference_accesses(var_accesses)
+    assert str(var_accesses) == "a: READ, b: READ, i: READ, p: WRITE, "\
+                                "q: READWRITE, r: READ"
+    # Test that the two accesses to 'q' indeed show up as
+    q_accesses = var_accesses["q"].all_accesses
+    assert len(q_accesses) == 2
+    assert q_accesses[0].access_type == AccessType.READ
+    assert q_accesses[1].access_type == AccessType.WRITE
+    assert q_accesses[0].location < q_accesses[1].location
+
+
+@pytest.mark.xfail(reason="Calls in the NEMO API are not yet supported #446")
+def test_call(parser):
+    ''' Check that we correctly handle a call in a program '''
+    reader = FortranStringReader('''program test_prog
+                                 call sub(a,b)
+                                 end program test_prog''')
+    ast = parser(reader)
+    psy = PSyFactory(API).create(ast)
+    schedule = psy.invokes.get("test_prog").schedule
+
+    code_block = schedule.children[0]
+    call_stmt = code_block.statements[0]
+    var_accesses = VariablesAccessInfo()
+    call_stmt.reference_accesses(var_accesses)
+    assert str(var_accesses) == "a: UNKNOWN, b: UNKNOWN"
+
+
+def test_do_loop(parser):
+    ''' Check the handling of do loops.
+    TODO #400: this only tests a nemo do loop. At this stage the loop
+    boundaries in nemo are only strings (not instances of Reference or so),
+    so in case of a loop like: "do jj=1, n", `n` is not be listed as
+    a READ access.
```

Comment:
Thanks for clarifying. I think this test should therefore be altered to check that `n` *is* listed as a READ access and then marked as x-fail. Ah, but then I see it would just be a duplicate of the test below. Is there a reason for having the two separate tests?

Reply:
Well, I think it's a good idea to have a passing test - it might check more of the code (e.g. taking a different branch in the code in case of the x-fail problem). I've marked the passing test with #400 and that it can be removed once the other (currently x-failing) test works as expected.

Reply:
I actually added more 'duplicated' tests - main problem is that if a test is marked as xfail, we can't distinguish the reason for the fail anymore. And with only one xfailing test for (say) gocean, any typo etc in gocean code that should be detect during testing would go unnoticed, since the tests would still xfail (and yes, that exactly happened to me ... I had a typo in my new is_scalar function, but no indication of any unexpected test failures 😢  ) . So I added a duplicated test for gocean and dynamo that is not xfailing, with enough comments to indicate that this should be removed once the xfailing tests work correctly.

Reply:
All sounds good to me - thanks for explaining.